hypertrace-gradle-publish-maven-central-plugin/src/main/java/org/hypertrace/gradle/publishing/PublishMavenCentralPlugin.java

```diff
@@ -0,0 +1,196 @@
+package org.hypertrace.gradle.publishing;
+
+import org.gradle.api.GradleException;
+import org.gradle.api.JavaVersion;
+import org.gradle.api.Plugin;
+import org.gradle.api.Project;
+import org.gradle.api.plugins.JavaPluginExtension;
+import org.gradle.api.publish.PublicationContainer;
+import org.gradle.api.publish.PublishingExtension;
+import org.gradle.api.publish.maven.MavenPublication;
+import org.gradle.api.publish.maven.plugins.MavenPublishPlugin;
+import org.gradle.api.tasks.javadoc.Javadoc;
+import org.gradle.external.javadoc.StandardJavadocDocletOptions;
+import org.gradle.plugins.signing.SigningExtension;
+import org.gradle.plugins.signing.SigningPlugin;
+
+import javax.annotation.Nonnull;
+
+public class PublishMavenCentralPlugin implements Plugin<Project> {
+  private static final String EXTENSION_NAME = "hypertracePublishMavenCentral";
+
+  private static final String PROPERTY_SIGNING_KEY_ID = "signingKeyId";
+  private static final String PROPERTY_SIGNING_KEY = "signingKey";
+  private static final String PROPERTY_SIGNING_PASSWORD = "signingPassword";
+  private static final String PROPERTY_OSSRH_USERNAME = "ossrhUsername";
+  private static final String PROPERTY_OSSRH_PASSWORD = "ossrhPassword";
+
+  private Project project;
+  private HypertracePublishMavenCentralExtension extension;
+
+  @Override
+  public void apply(@Nonnull Project target) {
+    project = target;
+    this.extension =
+      project.getExtensions().create(EXTENSION_NAME, HypertracePublishMavenCentralExtension.class);
+    this.applyMavenPublish();
+    this.applySigning();
+    this.applyWithJavadocJar();
+    this.applyWithSourcesJar();
+    this.addPublishRepository();
+    this.addPublications();
+  }
+
+  private void applySigning() {
+    project.getPluginManager()
+      .apply(SigningPlugin.class);
+  }
+
+  private void applyMavenPublish() {
+    project.getPluginManager()
+      .apply(MavenPublishPlugin.class);
+  }
+
+  private void applyWithJavadocJar() {
+    getJavaPluginExtension().withJavadocJar();
+    project.getTasks().withType(Javadoc.class).configureEach(javadoc -> {
+      StandardJavadocDocletOptions options = (StandardJavadocDocletOptions) javadoc.getOptions();
+      if (JavaVersion.current().isJava9Compatible()) {
+        options.addBooleanOption("html5", true);
+      }
+    });
+  }
+
+  private void applyWithSourcesJar() {
+    getJavaPluginExtension().withSourcesJar();
+  }
+
+  private void addPublishRepository() {
+    String url;
+    String user = getPropertyOrThrow(project, PROPERTY_OSSRH_USERNAME);
+    String password = getPropertyOrThrow(project, PROPERTY_OSSRH_PASSWORD);
+
+    if (project.getVersion().toString().endsWith("SNAPSHOT")) {
```

Comment:
Do we want to publish snapshots (no idea) ?

Reply:
sonatype has separate repositories for snapshot and release. we don't publish snapshot as of now.

Reply:
`snapshot` repository is private repo and is only available to member of the project. once the artifacts are published to `staging` repository, it needs to be closed and released: https://central.sonatype.org/pages/releasing-the-deployment.html. This can be done either manually or using a gradle plugin: https://github.com/Codearte/gradle-nexus-staging-plugin/.

we can use both plugins together. the new plugin use existing `javaLibrary` (if available). Although, we should avoid it as I am not sure the sequence in which these plugins will be configured/applied.